### ISSUE TITLE
Struct lifetime parameter

### ISSUE THREAD

**Issue by remi-dupre** [joshua-maros/ouroboros#9]
Hi!

It appears that there is an issue when referencing an outer lifetime. For example, take this simple toy example which takes a string slice and keep a subslice (this example is quite dumb as `int_slice` could be of type `&'s str` which doesn't need self reference):

```rust
#[ouroboros::self_referencing]
struct SelfRef<'s> {
    ext_slice: &'s str,
    #[borrows(ext_slice)]
    int_slice: &'this str,
}

impl<'s> SelfRef<'s> {
    fn foo(from: &'s str) -> SelfRef<'s> {
        SelfRefBuilder {
            ext_slice: from,
            int_slice: &[1..from],
        }
    }
}
```

We get the following compiler error:

```
error[E0392]: parameter `'s` is never used
   --> src/osm_reader/osm_store.rs:255:16
    |
255 | struct SelfRef<'s> {
    |                ^^ unused parameter
    |
    = help: consider removing `'s`, referring to it in a field, or using a marker such as `std::marker::PhantomData`
```

**Comment by remi-dupre** [comment]
Oh, I figured out this is probably a specific case of chain referencing, I'll close this until I test the chain hack for my use case :)

**Comment by remi-dupre** [comment]
Nah, actually I get the same error with the chain hack:

```rust
#[ouroboros::self_referencing(chain_hack)]
struct SelfRef<'s> {
    ext_slice: Box<&'s str>,
    #[borrows(ext_slice)]
    int_slice: &'this str,
}
```

**Comment by joshua-maros** [comment]
Confirmed, it looks like there's a couple of problems with the generated code in this case. I'll work on fixing it.

**Comment by joshua-maros** [comment]
The first error that was encountered was due to some of the generated structs not containing the value of `ext_slice`, causing the `'s is not used` error. Fixing that causes another error, which is actually valid, as `'this` may outlive `'s`, so `int_slice` should not be allowed to borrow `ext_slice`. Unfortunately the error is quite messy but I'm not sure how to improve it. If you change the dependency in your actual code to be:
```toml
ouroboros = { git = "https://github.com/joshua-maros/ouroboros" }
```
to test the new fixes I have added, do you encounter any additional errors?

**Comment by joshua-maros** [comment]
I have published the changes as version `0.4.1`.

**Comment by remi-dupre** [comment]
Awesome, thanks! I'll have a look into it :)

**Comment by wycats** [comment]
@joshua-maros I hit a similar issue, and was wondering whether it would be possible to allow users to declare `'this`:

```rust
#[ouroboros::self_referencing(chain_hack)]
struct SelfRef<'s, 'this> where 's: 'this {
    ext_slice: Box<&'s str>,
    #[borrows(ext_slice)]
    int_slice: &'this str,
}
```

**Comment by joshua-maros** [comment]
I did some testing and your comment that all lifetimes outlive `'this` appears to be correct, I was wrong in my initial assumption. Just to make sure, I've put a [post on the Rustlang forum](https://users.rust-lang.org/t/is-it-possible-for-a-struct-to-outlive-its-lifetime-parameter/51917) to get some input to make sure your statement is true in all cases. If so, I will make the macro automatically add conditions to the `where` clause such that all lifetime parameters outlive `this`.

**Comment by joshua-maros** [comment]
It looks like the consensus is that adding the automatic bounds would be safe. My schedule has been busy recently, but I should have time this weekend to add it in.

**Comment by wycats** [comment]
@joshua-maros that's really exciting. It will unlock some use-cases that I was hoping to use ouroboros for but couldn't quite get to work. Thanks!

**Comment by joshua-maros** [comment]
Turns out I didn't completely think it through. Since `'this` is converted to `'static` by the macro, it would be like putting `'a: 'static` on every lifetime which isn't useful. I'll comment again when I find a solution.

**Comment by joshua-maros** [comment]
It looks like there might not actually be a solution in the general case. I'm testing things out with this struct:
```rust
#[ouroboros::self_referencing]
struct Test<'a> {
    external: &'a String,
    #[borrows(external)]
    internal: &'this String,
}
```
Which, with the current version of the library, gets converted to:
```rust
struct Test<'a> {
    internal: &'static String,
    external: &'a String,
}
```
Which highlights the `'static` problem mentioned before. One option is to convert `&'this` to `*const` and have the generated code transparently convert between the two:
```rust
struct Test<'a> {
    internal: *const String,
    external
```
However, it wouldn't be able to handle the case where `'this` is used as a lifetime parameter for something other than a reference:
```rust
struct StringRef<'s>(&'s String);
#[ouroboros::self_referencing]
struct Test<'a> {
    external: StringRef<'a>,
    // No &'this to convert to *const, so we have to replace 'this with 'static
    internal: StringRef<'this>,
}
```
@wycats I would appreciate it if you could look over this and see if there is another way to make this work. Right now I can't think of any approach which would work consistently.

**Comment by tomprince** [comment]
> Turns out I didn't completely think it through. Since `'this` is converted to `'static` by the macro, it would be like putting `'a: 'static` on every lifetime which isn't useful. I'll comment again when I find a solution.

Would it be possible to use `'a` instead of `'static` in the case there is a single lifetime parameter?

**Comment by joshua-maros** [comment]
It would not work to automatically do that conversion, for example this would not compile correctly:
```rust
#[ouroboros::self_referencing]
struct Plain<'a> {
    data: String,
    #[borrows(data)]
    internal: &'this String,
    external: &'a String,
}
```
Doing it manually for this trivial example can be done without using this crate, but I anticipate this does not apply to the use cases other people have:
```rust
struct TwoRef<'a> {
    external: &'a String,
    internal: &'a String,
}

let mut value: TwoRef = ...;
value.internal = value.external;
```

**Comment by tomprince** [comment]
> It would not work to automatically do that conversion, for example this would not compile correctly: [...]

I'm not sure why it wouldn't? I manged (as an experiemnt) to get ouroboros to compile replacing `'static` with `'a` everywhere (so specialized to my particular use case) to test. With that, I could get everything to compile, but unfortunately https://github.com/rust-lang/rfcs/issues/1409 made any not trivial uses fail.

**Comment by tomprince** [comment]
> [...] but unfortunately [rust-lang/rfcs#1409](https://github.com/rust-lang/rfcs/issues/1409) made any not trivial uses fail.

Actually, I was mistaken, and my difficulty was related to the error type use in try_build. The code (which only compiles for a struct with a single lifetime parameter `'a`) is [here](https://github.com/joshua-maros/ouroboros/compare/master...tomprince:struct-lifetime). I've definitely not analyzed it for soundness, though.

**Comment by joshua-maros** [comment]
@tomprince After thinking about it more, this should be safe. The current approach is to replace `'this` with `'static` and then have functions that only provide access to these with a limited lifetime. Since any lifetime parameter will outlive the struct itself, it would be safe to use them in place of `'static` without any other changes. I've fleshed out the idea so that it will use the first lifetime parameter listed, and if there are no lifetime parameters it will use `'static` instead. It appears to be safe from what I've tested and it also seems to work with multiple lifetimes as long as the other lifetimes all have bounds like `'other: 'first`. It's now published as version `0.8.3`
